src/server/pkg/hashtree/db.go

```diff
@@ -0,0 +1,844 @@
+}
+
+func (h *dbHashTree) Serialize(w io.Writer) error {
+	return h.View(func(tx *bolt.Tx) error {
+		_, err := tx.WriteTo(w)
+		return err
+	})
+}
+
+func (h *dbHashTree) Copy() (HashTree, error) {
+	if err := h.Hash(); err != nil {
+		return nil, err
+	}
+	r, w := io.Pipe()
+	var eg errgroup.Group
+	eg.Go(func() (retErr error) {
+		defer func() {
+			if err := w.Close(); err != nil && retErr == nil {
+				retErr = err
+			}
+		}()
+		return h.Serialize(w)
+	})
+	var result HashTree
+	eg.Go(func() error {
+		var err error
+		result, err = DeserializeDBHashTree(r)
+		return err
+	})
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+func (h *dbHashTree) Deserialize(r io.Reader) error {
+	path := h.Path()
+	if err := h.Close(); err != nil {
+		return err
+	}
+	f, err := os.OpenFile(path, os.O_WRONLY|os.O_TRUNC, perm)
+	if err != nil {
+		return err
+	}
+	if _, err := io.Copy(f, r); err != nil {
+		return err
+	}
+	db, err := bolt.Open(path, perm, nil)
+	if err != nil {
+		return err
+	}
+	h.DB = db
+	return nil
+}
+
+func (h *dbHashTree) put(tx *bolt.Tx, path string, node *NodeProto) error {
+	data, err := node.Marshal()
+	if err != nil {
+		return err
+	}
+	if err := changed(tx).Put(b(path), changedVal); err != nil {
+		return fmt.Errorf("error putting \"%s\": %v", path, err)
+	}
+	return fs(tx).Put(b(path), data)
+}
+
+// visit visits every ancestor of 'path' (excluding 'path' itself), leaf to
+// root (i.e.  end of 'path' to beginning), and calls 'update' on each node
+// along the way. For example, if 'visit' is called with 'path'="/path/to/file",
+// then updateFn is called as follows:
+//
+// 1. update(node at "/path/to" or nil, "/path/to", "file")
+// 2. update(node at "/path"    or nil, "/path",    "to")
+// 3. update(node at "/"        or nil, "",         "path")
+//
+// This is useful for propagating changes to size upwards.
+func (h *dbHashTree) visit(tx *bolt.Tx, path string, update updateFn) error {
+	for path != "" {
+		parent, child := split(path)
+		pnode, err := h.get(tx, parent)
+		if err != nil && Code(err) != PathNotFound {
+			return err
+		}
+		if pnode != nil && pnode.nodetype() != directory {
+			return errorf(PathConflict, "attempted to visit \"%s\", but it's not a "+
+				"directory", path)
+		}
+		if pnode == nil {
+			pnode = &NodeProto{}
+		}
+		if err := update(pnode, parent, child); err != nil {
+			return err
+		}
+		if err := h.put(tx, parent, pnode); err != nil {
+			return err
+		}
+		path = parent
+	}
+	return nil
+}
+
+func (h *dbHashTree) PutFile(path string, objects []*pfs.Object, size int64) error {
+	return h.putFile(path, objects, nil, size)
+}
+
+func (h *dbHashTree) PutFileOverwrite(path string, objects []*pfs.Object, overwriteIndex *pfs.OverwriteIndex, sizeDelta int64) error {
+	return h.putFile(path, objects, overwriteIndex, sizeDelta)
+}
+
+func (h *dbHashTree) putFile(path string, objects []*pfs.Object, overwriteIndex *pfs.OverwriteIndex, sizeDelta int64) error {
```

Comment:
Just a note that I will need some changes here to store the BlockRefs.

Reply:
:+1: